File: backend/src/Solitaire.Engine/Pyramid/Pyramid.cs
using System.Collections.Immutable;

namespace Solitaire.Engine;

/// <summary>
/// The authoritative, pure Pyramid engine: deal, move validation/application,
/// legal-move generation, win detection, and deterministic replay. No method
/// mutates its inputs; every change yields a new <see cref="PyramidState"/>.
/// </summary>
/// <remarks>
/// Rules (per project decision): redeals are unlimited; the game is won once
/// the 28-card triangle is fully cleared (the stock/waste need not empty).
/// Pairs summing to 13 are removed together; a lone King (rank 13) is removed
/// by itself. Only the pyramid's currently-exposed cards and the single waste
/// top card are ever eligible — the stock is always face-down/inaccessible.
/// </remarks>
public static class Pyramid
{
    /// <summary>Creates the initial game state for a seed.</summary>
    public static PyramidState NewGame(int seed)
    {
        var shuffled = Deck.Shuffle(seed);
        var pyramid = ImmutableArray.CreateBuilder<Card?>(PyramidState.PyramidSize);
        int next = 0;
        for (int row = 0; row < PyramidState.RowCount; row++)
        {
            for (int col = 0; col <= row; col++)
            {
                pyramid.Add(shuffled[next++]);
            }
        }

        var stock = ImmutableArray.CreateBuilder<Card>(Deck.Size - next);
        for (int i = next; i < Deck.Size; i++)
        {
            stock.Add(shuffled[i]);
        }

        return new PyramidState(
            pyramid.MoveToImmutable(),
            stock.MoveToImmutable(),
            waste: ImmutableArray<Card>.Empty,
            score: 0);
    }

    /// <summary>True once the game is won (the whole triangle is cleared).</summary>
    public static bool IsWin(PyramidState state) => state.IsWon;

    /// <summary>
    /// Validates <paramref name="move"/> against the rules of
    /// <paramref name="state"/>. On success returns true and yields the resulting
    /// <paramref name="next"/> state and the nominal <paramref name="scoreDelta"/>.
    /// On failure returns false and makes no changes.
    /// </summary>
    public static bool TryApplyMove(PyramidState state, PyramidMove move, out PyramidState next, out int scoreDelta)
    {
        ArgumentNullException.ThrowIfNull(state);

        var result = Apply(state, move);
        if (result is null)
        {
            next = state;
            scoreDelta = 0;
            return false;
        }

        var (partial, delta) = result.Value;
        scoreDelta = delta;
        next = partial.With(score: state.Score + delta);
        return true;
    }

    /// <summary>Returns every legal move from <paramref name="state"/> (used for hints).</summary>
    public static IReadOnlyList<PyramidMove> GetLegalMoves(PyramidState state)
    {
        ArgumentNullException.ThrowIfNull(state);

        var moves = new List<PyramidMove>();

        if (!state.Stock.IsEmpty)
        {
            moves.Add(PyramidMove.Draw());
        }
        else if (!state.Waste.IsEmpty)
        {
            moves.Add(PyramidMove.Recycle());
        }

        var exposed = new List<int>();
        for (int i = 0; i < PyramidState.PyramidSize; i++)
        {
            if (IsExposed(state, i))
            {
                exposed.Add(i);
            }
        }

        var wasteTop = state.WasteTop;

        // Lone Kings.
        foreach (int i in exposed)
        {
            if (state.Pyramid[i]!.Value.Rank == Card.King)
            {
                moves.Add(PyramidMove.RemoveSingle(i));
            }
        }

        if (wasteTop is { Rank: Card.King })
        {
            moves.Add(PyramidMove.RemoveSingle(PyramidMove.Waste));
        }

        // Pairs among exposed pyramid cards.
        for (int a = 0; a < exposed.Count; a++)
        {
            int rankA = state.Pyramid[exposed[a]]!.Value.Rank;
            if (rankA == Card.King)
            {
                continue;
            }

            for (int b = a + 1; b < exposed.Count; b++)
            {
                int rankB = state.Pyramid[exposed[b]]!.Value.Rank;
                if (rankA + rankB == 13)
                {
                    moves.Add(PyramidMove.RemovePair(exposed[a], exposed[b]));
                }
            }
        }

        // Pairs between an exposed pyramid card and the waste top.
        if (wasteTop is { } waste && waste.Rank != Card.King)
        {
            foreach (int i in exposed)
            {
                int rank = state.Pyramid[i]!.Value.Rank;
                if (rank != Card.King && rank + waste.Rank == 13)
                {
                    moves.Add(PyramidMove.RemovePair(i, PyramidMove.Waste));
                }
            }
        }

        return moves;
    }

    /// <summary>
    /// Replays a move sequence from a fresh game. Pure and side-effect free: the
    /// result reports whether every move was legal, the final state and score,
    /// and whether the game was completed as a win.
    /// </summary>
    public static ReplayResult<PyramidState> Replay(int seed, IReadOnlyList<PyramidMove> moves)
    {
        ArgumentNullException.ThrowIfNull(moves);

        var state = NewGame(seed);

        for (int i = 0; i < moves.Count; i++)
        {
            if (!TryApplyMove(state, moves[i], out var nextState, out _))
            {
                return new ReplayResult<PyramidState>(
                    FinalState: state,
                    Score: state.Score,
                    Won: false,
                    AllMovesLegal: false,
                    FirstIllegalMoveIndex: i);
            }

            state = nextState;
        }

        return new ReplayResult<PyramidState>(
            FinalState: state,
            Score: state.Score,
            Won: state.IsWon,
            AllMovesLegal: true,
            FirstIllegalMoveIndex: null);
    }

    // ----------------------------------------------------------------------
    // Portable move (de)serialization for vectors and the API.
    // ----------------------------------------------------------------------

    public static MoveDto ToDto(PyramidMove move) => move.Type switch
    {
        PyramidMoveType.Draw => new MoveDto("Draw"),
        PyramidMoveType.Recycle => new MoveDto("Recycle"),
        PyramidMoveType.RemoveSingle => new MoveDto("RemoveSingle", Source: move.PositionA),
        PyramidMoveType.RemovePair =>
            new MoveDto("RemovePair", Source: move.PositionA, Destination: move.PositionB),
        _ => throw new ArgumentOutOfRangeException(nameof(move), move.Type, "Unknown move type."),
    };

    public static PyramidMove FromDto(MoveDto dto)
    {
        ArgumentNullException.ThrowIfNull(dto);
        return dto.Type switch
        {
            "Draw" => PyramidMove.Draw(),
            "Recycle" => PyramidMove.Recycle(),
            "RemoveSingle" => PyramidMove.RemoveSingle(Req(dto.Source, "source")),
            "RemovePair" =>
                PyramidMove.RemovePair(Req(dto.Source, "source"), Req(dto.Destination, "destination")),
            _ => throw new ArgumentOutOfRangeException(nameof(dto), dto.Type, "Unknown move type."),
        };
    }

    private static int Req(int? value, string field) =>
        value ?? throw new ArgumentException($"Pyramid move is missing '{field}'.");

    /// <summary>Pyramid has no options; the bag is accepted but ignored (parity with other variants).</summary>
    public static PyramidOptions OptionsFromBag(IReadOnlyDictionary<string, int> options)
    {
        ArgumentNullException.ThrowIfNull(options);
        return default;
    }

    // ----------------------------------------------------------------------
    // Rule helpers
    // ----------------------------------------------------------------------

    /// <summary>Flat index of row <paramref name="row"/>, column <paramref name="col"/> (both 0-based).</summary>
    private static int FlatIndex(int row, int col) => row * (row + 1) / 2 + col;

    /// <summary>The row (0..6) a flat pyramid index belongs to.</summary>
    private static int RowOf(int index)
    {
        int row = 0;
        while (FlatIndex(row + 1, 0) <= index)
        {
            row++;
        }

        return row;
    }

    /// <summary>
    /// True if the pyramid slot at <paramref name="index"/> holds a card AND
    /// both of the cards it rests on in the row below (its "children") are gone
    /// — or it is in the base row, which has nothing beneath it.
    /// </summary>
    private static bool IsExposed(PyramidState state, int index)
    {
        if (state.Pyramid[index] is null)
        {
            return false;
        }

        int row = RowOf(index);
        if (row == PyramidState.RowCount - 1)
        {
            return true;
        }

        int col = index - FlatIndex(row, 0);
        int childLeft = FlatIndex(row + 1, col);
        int childRight = FlatIndex(row + 1, col + 1);
        return state.Pyramid[childLeft] is null && state.Pyramid[childRight] is null;
    }

    /// <summary>Resolves a move position (pyramid slot or <see cref="PyramidMove.Waste"/>) to a card, if exposed/available.</summary>
    private static Card? ResolveExposedCard(PyramidState state, int position)
    {
        if (position == PyramidMove.Waste)
        {
            return state.WasteTop;
        }

        if (position < 0 || position >= PyramidState.PyramidSize)
        {
            return null;
        }

        return IsExposed(state, position) ? state.Pyramid[position] : null;
    }

    /// <summary>Removes the card at <paramref name="position"/> from wherever it lives.</summary>
    private static PyramidState RemoveAt(PyramidState state, int position)
    {
        if (position == PyramidMove.Waste)
        {
            return state.With(waste: state.Waste.RemoveAt(state.Waste.Length - 1));
        }

        return state.With(pyramid: state.Pyramid.SetItem(position, null));
    }

    // ----------------------------------------------------------------------
    // Move application. Returns null on any rule violation. Never mutates input.
    // ----------------------------------------------------------------------

    private static (PyramidState State, int ScoreDelta)? Apply(PyramidState state, PyramidMove move) => move.Type switch
    {
        PyramidMoveType.Draw => ApplyDraw(state),
        PyramidMoveType.Recycle => ApplyRecycle(state),
        PyramidMoveType.RemoveSingle => ApplyRemoveSingle(state, move),
        PyramidMoveType.RemovePair => ApplyRemovePair(state, move),
        _ => null,
    };

    private static (PyramidState, int)? ApplyDraw(PyramidState state)
    {
        if (state.Stock.IsEmpty)
        {
            return null;
        }

        var card = state.Stock[0];
        var newStock = state.Stock.RemoveAt(0);
        var newWaste = state.Waste.Add(card);
        return (state.With(stock: newStock, waste: newWaste), 0);
    }

    private static (PyramidState, int)? ApplyRecycle(PyramidState state)
    {
        if (!state.Stock.IsEmpty || state.Waste.IsEmpty)
        {
            return null;
        }

        // Flipping the waste pile over restores the original draw order:
        // waste[0] (first drawn) becomes the new stock top.
        return (state.With(stock: state.Waste, waste: ImmutableArray<Card>.Empty), 0);
    }

    private static (PyramidState, int)? ApplyRemoveSingle(PyramidState state, PyramidMove move)
    {
        var card = ResolveExposedCard(state, move.PositionA);
        if (card is not { Rank: Card.King })
        {
            return null;
        }

        return (RemoveAt(state, move.PositionA), PyramidScoring.RemoveSingle);
    }

    private static (PyramidState, int)? ApplyRemovePair(PyramidState state, PyramidMove move)
    {
        if (move.PositionA == move.PositionB)
        {
            return null;
        }

        var cardA = ResolveExposedCard(state, move.PositionA);
        var cardB = ResolveExposedCard(state, move.PositionB);
        if (cardA is not { } a || cardB is not { } b || a.Rank == Card.King || b.Rank == Card.King)
        {
            return null;
        }

        if (a.Rank + b.Rank != 13)
        {
            return null;
        }

        // Remove the higher position first so removing a pyramid slot can never
        // shift the meaning of the other position (positions are stable indices
        // into a flat array, not a list, so this ordering is actually not
        // required for correctness — kept for clarity/robustness regardless).
        var afterA = RemoveAt(state, move.PositionA);
        var afterBoth = RemoveAt(afterA, move.PositionB);
        return (afterBoth, PyramidScoring.RemovePair);
    }
}

File: frontend/src/engine/pyramid.ts
// Pyramid engine — a behavioral port of the C# Pyramid/*.cs files. Same deal,
// same exposure/pairing rules, same scoring.
//
// Rules (per project decision): redeals are unlimited; the game is won once
// the 28-card triangle is fully cleared (the stock/waste need not empty).
// Pairs summing to 13 are removed together; a lone King (rank 13) is removed
// by itself. Only the pyramid's currently-exposed cards and the single waste
// top card are ever eligible — the stock is always face-down/inaccessible.

import { KING, type Card } from './cards';
import { DECK_SIZE, shuffle } from './deck';
import type {
  ApplyResult,
  GameDefinition,
  MoveDto,
  ReplayOutcome,
  ReplayResult,
  SolitaireEngine,
} from './types';

export const PYRAMID_ROW_COUNT = 7;
export const PYRAMID_SIZE = 28; // 1+2+...+7

/** Sentinel position meaning "the waste's top card" rather than a pyramid slot. */
export const PYRAMID_WASTE = -1;

const SCORE = {
  removePair: 15,
  removeSingle: 10,
} as const;

export interface PyramidState {
  readonly pyramid: readonly (Card | undefined)[]; // length 28, flat row-major; undefined = removed
  readonly stock: readonly Card[]; // index 0 = top (next drawn)
  readonly waste: readonly Card[]; // last element = playable top
  readonly score: number;
}

export function pyramidIsWon(state: PyramidState): boolean {
  return state.pyramid.every((card) => card === undefined);
}

export function pyramidNewGame(seed: number): PyramidState {
  const shuffled = shuffle(seed);
  const pyramid: (Card | undefined)[] = [];
  let next = 0;
  for (let row = 0; row < PYRAMID_ROW_COUNT; row++) {
    for (let col = 0; col <= row; col++) {
      pyramid.push(shuffled[next++]);
    }
  }
  const stock = shuffled.slice(next, DECK_SIZE);
  return { pyramid, stock, waste: [], score: 0 };
}

type Applied = { readonly state: PyramidState; readonly delta: number } | null;

export function pyramidTryApplyMove(
  state: PyramidState,
  move: MoveDto,
): ApplyResult<PyramidState> {
  const result = apply(state, move);
  if (result === null) {
    return { ok: false, next: state, scoreDelta: 0 };
  }
  return {
    ok: true,
    next: { ...result.state, score: state.score + result.delta },
    scoreDelta: result.delta,
  };
}

export function pyramidReplay(seed: number, moves: readonly MoveDto[]): ReplayResult<PyramidState> {
  let state = pyramidNewGame(seed);
  for (let i = 0; i < moves.length; i++) {
    const applied = pyramidTryApplyMove(state, moves[i]!);
    if (!applied.ok) {
      return {
        finalState: state,
        score: state.score,
        won: false,
        allMovesLegal: false,
        firstIllegalMoveIndex: i,
      };
    }
    state = applied.next;
  }
  return {
    finalState: state,
    score: state.score,
    won: pyramidIsWon(state),
    allMovesLegal: true,
    firstIllegalMoveIndex: null,
  };
}

export function pyramidGetLegalMoves(state: PyramidState): MoveDto[] {
  const moves: MoveDto[] = [];

  if (state.stock.length > 0) {
    moves.push({ type: 'Draw' });
  } else if (state.waste.length > 0) {
    moves.push({ type: 'Recycle' });
  }

  const exposed: number[] = [];
  for (let i = 0; i < PYRAMID_SIZE; i++) {
    if (isExposed(state, i)) {
      exposed.push(i);
    }
  }

  const wasteTop = pyramidWasteTop(state);

  for (const i of exposed) {
    if (state.pyramid[i]!.rank === KING) {
      moves.push({ type: 'RemoveSingle', source: i });
    }
  }
  if (wasteTop !== undefined && wasteTop.rank === KING) {
    moves.push({ type: 'RemoveSingle', source: PYRAMID_WASTE });
  }

  for (let a = 0; a < exposed.length; a++) {
    const rankA = state.pyramid[exposed[a]!]!.rank;
    if (rankA === KING) {
      continue;
    }
    for (let b = a + 1; b < exposed.length; b++) {
      const rankB = state.pyramid[exposed[b]!]!.rank;
      if (rankA + rankB === 13) {
        moves.push({ type: 'RemovePair', source: exposed[a]!, destination: exposed[b]! });
      }
    }
  }

  if (wasteTop !== undefined && wasteTop.rank !== KING) {
    for (const i of exposed) {
      const rank = state.pyramid[i]!.rank;
      if (rank !== KING && rank + wasteTop.rank === 13) {
        moves.push({ type: 'RemovePair', source: i, destination: PYRAMID_WASTE });
      }
    }
  }

  return moves;
}

export const pyramidEngine: SolitaireEngine = {
  variant: 'pyramid',
  replay(game: GameDefinition): ReplayOutcome {
    const result = pyramidReplay(game.seed, game.moves);
    return {
      score: result.score,
      won: result.won,
      allMovesLegal: result.allMovesLegal,
      firstIllegalMoveIndex: result.firstIllegalMoveIndex,
    };
  },
};

// -- rule helpers -------------------------------------------------------------

function pyramidWasteTop(state: PyramidState): Card | undefined {
  return state.waste.length === 0 ? undefined : state.waste[state.waste.length - 1];
}

/** Flat index of row `row`, column `col` (both 0-based). */
function flatIndex(row: number, col: number): number {
  return (row * (row + 1)) / 2 + col;
}

/** The row (0..6) a flat pyramid index belongs to. */
function rowOf(index: number): number {
  let row = 0;
  while (flatIndex(row + 1, 0) <= index) {
    row++;
  }
  return row;
}

/**
 * True if the pyramid slot at `index` holds a card AND both of the cards it
 * rests on in the row below (its "children") are gone — or it is in the base
 * row, which has nothing beneath it.
 */
function isExposed(state: PyramidState, index: number): boolean {
  if (state.pyramid[index] === undefined) {
    return false;
  }
  const row = rowOf(index);
  if (row === PYRAMID_ROW_COUNT - 1) {
    return true;
  }
  const col = index - flatIndex(row, 0);
  const childLeft = flatIndex(row + 1, col);
  const childRight = flatIndex(row + 1, col + 1);
  return state.pyramid[childLeft] === undefined && state.pyramid[childRight] === undefined;
}

/** Resolves a move position (pyramid slot or PYRAMID_WASTE) to a card, if exposed/available. */
function resolveExposedCard(state: PyramidState, position: number): Card | undefined {
  if (position === PYRAMID_WASTE) {
    return pyramidWasteTop(state);
  }
  if (position < 0 || position >= PYRAMID_SIZE) {
    return undefined;
  }
  return isExposed(state, position) ? state.pyramid[position] : undefined;
}

/** Removes the card at `position` from wherever it lives. */
function removeAt(state: PyramidState, position: number): PyramidState {
  if (position === PYRAMID_WASTE) {
    return { ...state, waste: state.waste.slice(0, -1) };
  }
  const pyramid = [...state.pyramid];
  pyramid[position] = undefined;
  return { ...state, pyramid };
}

// -- move application (returns null on any rule violation) --------------------

function apply(state: PyramidState, move: MoveDto): Applied {
  switch (move.type) {
    case 'Draw':
      return applyDraw(state);
    case 'Recycle':
      return applyRecycle(state);
    case 'RemoveSingle':
      return applyRemoveSingle(state, move.source ?? -2);
    case 'RemovePair':
      return applyRemovePair(state, move.source ?? -2, move.destination ?? -2);
    default:
      return null;
  }
}

function applyDraw(state: PyramidState): Applied {
  if (state.stock.length === 0) {
    return null;
  }
  const card = state.stock[0]!;
  return {
    state: { ...state, stock: state.stock.slice(1), waste: [...state.waste, card] },
    delta: 0,
  };
}

function applyRecycle(state: PyramidState): Applied {
  if (state.stock.length > 0 || state.waste.length === 0) {
    return null;
  }
  // Flipping the waste pile over restores the original draw order: waste[0]
  // (first drawn) becomes the new stock top.
  return { state: { ...state, stock: [...state.waste], waste: [] }, delta: 0 };
}

function applyRemoveSingle(state: PyramidState, position: number): Applied {
  const card = resolveExposedCard(state, position);
  if (card === undefined || card.rank !== KING) {
    return null;
  }
  return { state: removeAt(state, position), delta: SCORE.removeSingle };
}

function applyRemovePair(state: PyramidState, positionA: number, positionB: number): Applied {
  if (positionA === positionB) {
    return null;
  }
  const a = resolveExposedCard(state, positionA);
  const b = resolveExposedCard(state, positionB);
  if (a === undefined || b === undefined || a.rank === KING || b.rank === KING) {
    return null;
  }
  if (a.rank + b.rank !== 13) {
    return null;
  }
  const afterA = removeAt(state, positionA);
  const afterBoth = removeAt(afterA, positionB);
  return { state: afterBoth, delta: SCORE.removePair };
}
